src/Microsoft.AspNetCore.OData/Routing/ODataActionSelector.cs

```diff
@@ -160,12 +190,27 @@ private bool TryMatch(IList<ParameterDescriptor> parameters, IList<string> avail
                     }
                 }
 
+                // if parameter is not bound to a key in the path,
+                // assume that it's bound to the request body
+                // only one parameter should be considered bound to the body
+                if (!matchedBody && RequestHasBody(context))
+                {
+                    matchedBody = true;
+                    continue;
+                }
+
                 return false;
             }
 
             return true;
         }
 
+        private bool RequestHasBody(RouteContext context)
+        {
+            return context.HttpContext.Request.ContentLength > 0 ||
```

Comment:
>context.HttpContext.Request.ContentLength > 0 [](start = 19, length = 45)

this change seems to require the setting of content length ahead? It maybe a regression? what can we find a workaround for it?

Reply:
My reasoning was that all real-world requests coming from a compliant http client would include the content-length header for requests with a body since it's required by the [HTTP spec](https://www.w3.org/Protocols/HTTP/1.0/draft-ietf-http-spec.html#Content-Length) and so I assume it would be safe to rely on it. Initially I had tried to get the length directly from the stream but this was not reliable cause the stream is not guaranteed to be seekable and didn't work consistently between .net framework and .net core.

I also updated tests that create requests with bodies to include the content-length header. I don't know of a better workaround.